File: src/main/java/client/inventory/ModifyInventoriesContext.java
package client.inventory;

import client.inventory.item.templates.ItemTemplate;
import client.inventory.operations.AbstractModifyInventoryOperation;
import client.inventory.slots.ItemSlot;
import client.inventory.slots.ItemSlotBundle;
import util.packet.PacketWriter;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ModifyInventoriesContext {

    private final Map<ItemInventoryType, ModifyInventoryContext> inventories = new HashMap<>();

    public ModifyInventoriesContext(Map<ItemInventoryType, ItemInventory> inventories) {
        for (Map.Entry<ItemInventoryType, ItemInventory> kv : inventories.entrySet()) {
            this.inventories.put(kv.getKey(), new ModifyInventoryContext(kv.getKey(), kv.getValue()));
        }
    }

    public List<AbstractModifyInventoryOperation> getOperations() {
        List<AbstractModifyInventoryOperation> operations = new ArrayList<>();
        inventories.values().stream().map(ModifyInventoryContext::getOperations).forEach(operations::addAll);

        return operations;
    }

    public ModifyInventoryContextInterface getInventoryContext(ItemInventoryType type) {
        return inventories.get(type);
    }

    public ModifyInventoryContextInterface getInventoryByItemId(int id) {
        return getInventoryContext(ItemInventoryType.values()[(id / 1000000) - 1]);
    }

    public void add(ItemSlot item) {
        getInventoryByItemId(item.getTemplateId()).add(item);
    }

    public void add(ItemTemplate item, short quantity) {
        getInventoryByItemId(item.getId()).add(item, quantity);
    }

    public void set(short slot, ItemSlot item) {
        getInventoryByItemId(item.getTemplateId()).set(slot, item);
    }

    public void set(short slot, ItemTemplate item, short quantity) {
        getInventoryByItemId(item.getId()).set(slot, item, quantity);
    }

    public void remove(ItemSlot item) {
        getInventoryByItemId(item.getTemplateId()).remove(item);
    }

    public void remove(ItemSlot item, short quantity) {
        getInventoryByItemId(item.getTemplateId()).remove(item, quantity);
    }

    public void remove(int id, short quantity) {
        getInventoryByItemId(id).remove(id, quantity);
    }

    public ItemSlotBundle take(ItemSlotBundle bundle, short quantity) {
        return getInventoryByItemId(bundle.getTemplateId()).take(bundle, quantity);
    }

    public ItemSlotBundle take(int id, short quantity) {
        return getInventoryByItemId(id).take(id, quantity);
    }

    public void update(ItemSlot item) {
        getInventoryByItemId(item.getTemplateId()).update(item);
    }

    public void encode(PacketWriter pw) {
        List<AbstractModifyInventoryOperation> operations = getOperations();
        pw.writeByte((byte) operations.size());
        operations.forEach(op -> op.encode(pw));
    }
}

File: src/main/java/client/inventory/ModifyInventoriesContext.kt
package client.inventory

import client.inventory.item.slots.ItemSlot
import client.inventory.item.slots.ItemSlotBundle
import client.inventory.item.templates.ItemTemplate
import client.inventory.operations.AbstractModifyInventoryOperation
import util.packet.PacketWriter
import java.util.*

class ModifyInventoriesContext(inventories: Map<ItemInventoryType, ItemInventory>) {

    private val inventories: MutableMap<ItemInventoryType, ModifyInventoryContext> = EnumMap(ItemInventoryType::class.java)

    val operations: List<AbstractModifyInventoryOperation>
        get() {
            val operations: MutableList<AbstractModifyInventoryOperation> = ArrayList()
            inventories.values.stream().map { it.operations }.forEach { operations.addAll(it) }
            return operations
        }

    fun getInventoryContext(type: ItemInventoryType): ModifyInventoryContextInterface {
        return inventories[type]!!
    }

    fun getInventoryByItemId(id: Int): ModifyInventoryContextInterface {
        return getInventoryContext(ItemInventoryType.values()[id / 1000000 - 1])
    }

    fun add(item: ItemSlot) {
        getInventoryByItemId(item.templateId).add(item)
    }

    fun add(item: ItemTemplate, quantity: Short) {
        getInventoryByItemId(item.id).add(item, quantity)
    }

    operator fun set(slot: Short, item: ItemSlot) {
        getInventoryByItemId(item.templateId)[slot] = item
    }

    operator fun set(slot: Short, item: ItemTemplate, quantity: Short) {
        getInventoryByItemId(item.id)[slot, item] = quantity
    }

    fun remove(item: ItemSlot) {
        getInventoryByItemId(item.templateId).remove(item)
    }

    fun remove(item: ItemSlot, quantity: Short) {
        getInventoryByItemId(item.templateId).remove(item, quantity)
    }

    fun remove(id: Int, quantity: Short) {
        getInventoryByItemId(id).remove(id, quantity)
    }

    fun take(bundle: ItemSlotBundle, quantity: Short): ItemSlotBundle {
        return getInventoryByItemId(bundle.templateId).take(bundle, quantity)
    }

    fun take(id: Int, quantity: Short): ItemSlotBundle {
        return getInventoryByItemId(id).take(id, quantity)
    }

    fun update(item: ItemSlot) {
        getInventoryByItemId(item.templateId).update(item)
    }

    fun encode(pw: PacketWriter) {
        val operations = operations
        pw.writeByte(operations.size.toByte())
        operations.forEach { it.encode(pw) }
    }

    init {
        for ((key, value) in inventories) {
            this.inventories[key] = ModifyInventoryContext(key, value)
        }
    }
}